src/Cli/Microsoft.DotNet.FileBasedPrograms/FileLevelDirectiveHelpers.cs

```diff
@@ -0,0 +1,623 @@
+            }
+
+            return new Project(context.Info)
+            {
+                Name = directiveText,
+            };
+        }
+
+        public Project WithName(string name)
+        {
+            return new Project(Info) { Name = name };
+        }
+
+        public static FileInfo GetProjectFileFromDirectory(string projectDirectory)
+        {
+            DirectoryInfo dir;
+            try
+            {
+                dir = new DirectoryInfo(projectDirectory);
+            }
+            catch (ArgumentException)
+            {
+                throw new GracefulException(FileBasedProgramsResources.CouldNotFindProjectOrDirectory, projectDirectory);
+            }
+
+            if (!dir.Exists)
+            {
+                throw new GracefulException(FileBasedProgramsResources.CouldNotFindProjectOrDirectory, projectDirectory);
+            }
+
+            FileInfo[] files = dir.GetFiles("*proj");
+            if (files.Length == 0)
+            {
+                throw new GracefulException(
+                    FileBasedProgramsResources.CouldNotFindAnyProjectInDirectory,
+                    projectDirectory);
+            }
+
+            if (files.Length > 1)
+            {
+                throw new GracefulException(FileBasedProgramsResources.MoreThanOneProjectInDirectory, projectDirectory);
+            }
+
+            return files.First();
+        }
+
+        public override string ToString() => $"#:project {Name}";
+    }
+}
+
+/// <summary>
+/// Used for deduplication - compares directives by their type and name (ignoring case).
+/// </summary>
+internal sealed class NamedDirectiveComparer : IEqualityComparer<CSharpDirective.Named>
+{
+    public static readonly NamedDirectiveComparer Instance = new();
+
+    private NamedDirectiveComparer() { }
+
+    public bool Equals(CSharpDirective.Named? x, CSharpDirective.Named? y)
+    {
+        if (ReferenceEquals(x, y)) return true;
+
+        if (x is null || y is null) return false;
+
+        return x.GetType() == y.GetType() &&
+            StringComparer.OrdinalIgnoreCase.Equals(x.Name, y.Name);
+    }
+
+    public int GetHashCode(CSharpDirective.Named obj)
+    {
+        return ExternalHelpers.CombineHashCodes(
+            obj.GetType().GetHashCode(),
+            StringComparer.OrdinalIgnoreCase.GetHashCode(obj.Name));
+    }
+}
+
+internal sealed class SimpleDiagnostic
+{
+    public required Position Location { get; init; }
+    public required string Message { get; init; }
+
+    /// <summary>
+    /// An adapter of <see cref="FileLinePositionSpan"/> that ensures we JSON-serialize only the necessary fields.
+    /// </summary>
+    public readonly struct Position
+    {
+        // note: this type is only serialized for run-api scenarios.
+        // If/when run-api is removed, we would also want to remove this configuration.
```

Comment:
This comment applies to the whole `Position` struct, right? Consider moving it to a `<remarks>` section there, as it currently seems like it would apply only to `TextSpan` property.